File: Residential_Controller.py
import math;
import time


# // **************************************************************************

# //                           BATTERY

# // **************************************************************************

numberOfFloors = 10

class Battery:
    def __init__(self, numberOfFloors, numberOfColumns):
        self.numberOfFloors = numberOfFloors
        self.columns = self.initColumns(numberOfFloors, numberOfColumns)
        print(str(self.columns));

    def initColumns(self, numberOfFloors, numberOfColumns):
        floorRange = math.floor(numberOfFloors / numberOfColumns)
        columns = []
        floor = 1
        maxFloor = numberOfFloors
        columnId = 0

        print(str(floorRange))

        while floor < maxFloor:
            if floor == 1:
                lowestFloor = 2
            else:
                lowestFloor = floor + 1
            

            if floor + floorRange > numberOfFloors:
                highestFloor = numberOfFloors
            else:
                highestFloor = floor + floorRange

            floor = floor + floorRange
            columnId = columnId + 1
            
            columns.append(Column(columnId, lowestFloor, highestFloor))

        return columns

    def getColumnButton(self, floor, direction):
        response = None

        for columnItem in self.columns:
            for floorButton in columnItem.floorButtons:
                print('floorButton ' + str(floorButton.floorNumber) + "  " +str(floorButton.direction))
                if floorButton.floorNumber == floor and floorButton.direction ==  direction:
                    response = {"column": columnItem, "button": floorButton}
                    
        
        print(str(response))
        return response




# // **************************************************************************

# //                           COLUMN

# // **************************************************************************

scene = [
    [{
        "currentFloor": 10,
        "tasksList": []
    }, {
        "currentFloor": 3,
        "tasksList": [6]
    }], [{
        "currentFloor": 10,
        "tasksList": []
    }, {
        "currentFloor": 3,
        "tasksList": [6]
    }]
]   

class Column:
    def __init__(self, columnId, lowestFloor, highestFloor):
        self.id = columnId
        self.floorButtons = self.initButtons(lowestFloor, highestFloor, columnId)
        self.elevators = self.initElevators(scene, columnId)
        self.lowestFloor = lowestFloor
        self.highestFloor = highestFloor
        print('new Column')

    def initButtons(self, lowestFloor, highestFloor, columnId):
        buttons = []

        floor = lowestFloor
        maxFloor = highestFloor

        buttons.insert(0, FloorButton('up', 1, self.id))

        for i in range(floor, highestFloor + 1):
            buttons.append(FloorButton('up', i, self.id));
            buttons.append(FloorButton('down', i, self.id))

        return buttons

    def initElevators(self, scene, columnId):
        elevatorsList = []
        index = 0
        currentScene = scene[columnId]

        for elevator in currentScene:
            elevatorsList.append(Elevator(elevator['currentFloor'], elevator['tasksList'], index))
            index = index + 1
    
        print(elevatorsList)
        return elevatorsList
    
    def requestElevator(self, floor, direction):
        print('Elevator request at : ' + str(floor) + " to go " + str(direction))

        request = ElevatorRequest(floor, direction)
        elevator = self.getBestElevator(request)

        time.sleep(1)
        print("Rocket Elevator hove found you an elevator | id: " + str(elevator.id) + "is comming from floor: " + str(elevator.currentFloor))
        elevator.addTask(floor)
        time.sleep(1)
        elevator.operate()
        
    def getBestElevator(self, request):
        idleElevators = []
        commingElevators = []
        othersElevators = []

        for elevator in self.elevators:
            if len(elevator.tasksList) == 0:
                idleElevators.append(elevator)
                print('elevator appended in idle')
            elif elevator.isComming(request):
                commingElevators.append(elevator)
                print('appende to comming')
            else: 
                othersElevators.append(elevator)
                print('appended to others')

        bestElevator = None
        
        if len(commingElevators) > 0:
            bestElevator = self.getBestFrom(commingElevators, request)
        elif len(idleElevators) > 0:
            bestElevator = self.getBestFrom(idleElevators, request)
        else:
            bestElevator = self.getBestFrom(othersElevators, request)

        return bestElevator

    def getBestFrom(self, list, request):
        bestElevator = None
        bestTravelSteps = None

        for elevator in list:
            elevatorTravelSteps = self.getStepsToCome(elevator, request)

            if bestTravelSteps == None or elevatorTravelSteps < bestTravelSteps:
                bestElevator = elevator
                bestTravelSteps = elevatorTravelSteps

        return bestElevator

    def getStepsToCome(self, elevator, request):
        print('getStepsToCome with : ' + str(elevator))
        direction = elevator.getDirection()
        floor = request.requestedFloor
        stepsToCome = 0
        previousTask = elevator.currentFloor

        if len(elevator.tasksList) == 0:
            stepsToCome = stepsToCome + abs(previousTask - floor)
        else:
            index = 0
            for task in elevator.tasksList:
                nextTask = None

                if elevator.tasksList[index]:
                    nextTask = elevator.tasksList[index]
                else:
                    nextTask = floor

                if direction == "up" and floor >= previousTask and floor <= nextTask:
                    stepsToCome = stepsToCome + abs(previousTask - floor);
                elif direction == "down" and floor <= previousTask and floor >= nextTask:
                    stepsToCome == stepsToCome + abs(previousTask - nextTask)
                else:
                    stepsToCome = stepsToCome + abs(previousTask - nextTask)

                previousTask = nextTask
        
        return stepsToCome
                    







class ElevatorRequest:
    def __init__(self, requestedFloor, direction):
        self.requestedFloor = requestedFloor
        self.direction = direction
        



# // **************************************************************************

# //      ELEVATOR

# // **************************************************************************

class Elevator:
    def __init__(self, currentFloor, tasksList, elevatorId):
        self.id = elevatorId
        self.currentFloor = currentFloor
        self.doorState = 'closed'
        self.isSafe = True
        self.maxWheit = 3500
        self.buttonsList = []
        self.tasksList = tasksList
        print("new Elevator: "+  str(self.id))

    def isComming(self, request):
        print('is comming')
        elevatorDirection = self.getDirection()
        isElevatorComming = False

        if elevatorDirection == request.direction:
            if elevatorDirection == "up":
                for task in self.tasksList:
                    if self.currentFloor <= request.requestedFloor and self.currentFloor < task:
                        isElevatorComming = True
                        break
            elif elevatorDirection == "down":
                for task in self.tasksList:
                    if self.currentFloor >= request.requestedFloor and self.currentFloor > task:
                        isElevatorComming = True
                        break
        
        print("isElevatorComming " + str(isElevatorComming))
        return isElevatorComming

    def getDirection(self):
        movingDown = None
        movingUp = None

        if len(self.tasksList) == 1 and self.currentFloor > self.tasksList[0]:
            movingDown = True
        if len(self.tasksList) > 2:
            if self.tasksList[len(self.tasksList) - 2] > self.tasksList[len(self.tasksList) - 1]:
                movingDown = True

        if len(self.tasksList) == 1 and self.currentFloor < self.tasksList[0]:
            movingUp = True
        if len(self.tasksList) > 2:
            if self.tasksList[len(self.tasksList) - 2] < self.tasksList[len(self.tasksList) - 1]:
                movingUp = True


        if len(self.tasksList) == 0:
            print('none')
            return None
        elif movingDown:
            print('down')
            return 'down'

        else:
            print('up')
            return 'up'

    def operate(self):
        print('operate')

        again = True
        if len(self.tasksList) > 0:
            time.sleep(1)
            while again:
                time.sleep(1)
                if self.tasksList[0] > self.currentFloor:
                    self.moveUp()
                    print('elevator is moving and at floor : ' + str(self.currentFloor))
                elif self.tasksList[0] < self.currentFloor:
                    self.moveDown()
                    print('elevator is moving and at floor requested: ' + str(self.currentFloor))
                else: 
                    self.openDoor()
                    self.removeTask()
                    # time(1)
                    self.closeDoor()
                    again = False

    def moveDown(self):
        self.currentFloor = self.currentFloor - 1
        print('move down')

    def moveUp(self):
        self.currentFloor = self.currentFloor + 1
        print('move up')

    def openDoor(self):
        print('elevator is at floor requested: ' + str(self.currentFloor))
        time.sleep(1)
        self.doorState = "opened"
        time.sleep(1)
        print('open doors')

    def closeDoor(self):
        self.doorState = "closed"
        print("waiting 5 seconds befor closing")
        time.sleep(5)
        print('close doors')

    def addTask(self, task):
        self.tasksList.append(task)
        print('add Task')
        print('YOUR ELEVATOR IS MOVING TO THE REQUESTED FLOOR ' +  str(task))

    def removeTask(self):
        del self.tasksList[0]
        print('remove task')






# // **************************************************************************

# //      FLOOR_BUTTON

# // **************************************************************************
class FloorButton:
    def __init__(self, direction, number, columnId):
        self.columnId = columnId
        self.floorNumber = number
        self.direction = direction
        self.light = False

    def requestElevator(self):
        # requestElevator(self.FloorNumber, self.direction)
        self.toggleLight()
        print('light')

    def toggleLight(self):
        self.light = not self.light
        print('light: ' + str(self.light))






battery = Battery(10,1)

# battery.columns[0].elevators[1].operate()
def requestElevator(floor, direction):
    print('requestElevator at floor: ' + str(floor) + " and direction: " + str(direction))

    query = battery.getColumnButton(floor, direction)

    if query == None:
        print("Your request cannot be handled, maybe the floor exists in your dreams !")
    else:
        query['column'].requestElevator(floor, direction)

    # if query:
    #     query.column.requestElevator(floor, direction)
    # else:
    #     print('the request you trying to make is not possible because only God can')



def requestFloor(elevatorId, requestedFloor):
    for col in battery.columns:
        print('col: ' + str(col))
        for elevatorItem in col.elevators:
            print( elevatorItem.id)
            if elevatorItem.id == elevatorId:
                print('id: ')
                elevatorItem.addTask(requestedFloor)
                elevatorItem.operate()
                print('Elevator found,' +  str(elevatorItem))
                print('YOUR ELEVATOR IS MOVING TO THE REQUESTED FLOOR ' +  str(requestedFloor))
            #     break

requestFloor(1, 7);

File: Residential_Controller.js
// // **************************************************************************

// //                           BATTERY

// // **************************************************************************

class Battery {
    constructor(numberOfFloors, numberOfColumns) {
        this.numberOfFloors = numberOfFloors ? numberOfFloors : 10
        this.columns = this.initColumns(numberOfFloors, numberOfColumns)
    }

    initColumns = (numberOfFloors, numberOfColumns) => {
        let range = Math.floor(numberOfFloors / numberOfColumns);
        let columns = []

        let floor = 1
        let maxFloor = numberOfFloors
        let columnId = 0
        for (; floor < maxFloor; floor += range) {
            let lowestFloorLevel = floor == 1 ? 2 : floor + 1;
            let highestFloorLevel = (floor + range) > numberOfFloors ? numberOfFloors : floor + range

            columns.push(new Column(columnId++, lowestFloorLevel, highestFloorLevel));
        }

        return columns
    }

    getColumnButton = (floor, direction) => {
        let response;

        battery.columns.forEach(columnItem => {
            columnItem.floorButtons.some(buttonItem => {
                if (buttonItem.floorNumber == floor && buttonItem.direction == direction) {

                    response = {
                        column: columnItem,
                        button: buttonItem
                    }

                }
            })
        })

        return response
    }
}









// **************************************************************************

//                              COLUMN

// **************************************************************************

let elevators = [
    [{
        currentFloor: 10,
        tasksList: []
    }, {
        currentFloor: 3,
        tasksList: [6]
    }]
]


class Column {
    constructor(id, lowestFloor, highestFloor) {
        this.id = id
        this.floorButtons = this.initButtons(lowestFloor, highestFloor, id)
        this.elevators = this.initElevators(elevators[id])
        console.log("Column: ", this.floorButtons, this.elevators);
    }

    initButtons = (lowestFloor, highestFloor, columnId) => {
        let buttons = []

        let floor = lowestFloor
        let maxFloor = highestFloor

        buttons.unshift(new FloorButton('up', 1, columnId))

        for (; floor <= maxFloor; floor++) {
            buttons.push(new FloorButton('up', floor, columnId))
            buttons.push(new FloorButton('down', floor, columnId));
        }

        return buttons
    }

    initElevators = (scene) => {
        let elevatorsList = []

        if (scene) scene.forEach((elevator, index) =>
            elevatorsList.push(new Elevator(elevator, index + 1))
        )

        return elevatorsList
    }

    requestElevator = (requestedFloor, direction) => {
        let elevatorRequest = new ElevatorRequest(requestedFloor, direction);
        let elevator = this.getBestElevator(elevatorRequest)
        
        console.log(requestedFloor)
        elevator.addTask(requestedFloor);
    }

    getBestElevator = (request) => {
        let idleElevators = []
        let commingElevators = []
        let othersElevators = []


        this.elevators.forEach(elevator => {
            if (elevator.tasksList == 0) idleElevators.push(elevator)
            else if (elevator.isComming(request)) commingElevators.push(elevator)
            else othersElevators.push(elevator)
        });


        let bestElevator;
        [commingElevators, idleElevators, othersElevators].some(list => {
             
            bestElevator = this.getFastestFrom(list, request)
            return bestElevator
        })
        return bestElevator
    }

    getFastestFrom = (elevatorsList, request) => {
        let bestElevator;
        let bestTravelSteps;

        console.log('getBestFrom: ', elevatorsList, request)
        elevatorsList.forEach(elevator => {
            let elevatorTravelSteps = this.getStepsToCome(elevator, request)

            if (bestTravelSteps == undefined || elevatorTravelSteps < bestTravelSteps) {
                bestElevator = elevator
                bestTravelSteps = elevatorTravelSteps
            }
        })
        return bestElevator
    }


    getStepsToCome = (elevator, request) => {
        let direction = elevator.getDirection()
        let floor = request.requestedFloor
        let idle = elevator.tasksList.length == 0
        let stepsToCome = 0
        let previousTask = elevator.currentFloor

         
        if (idle) {
            stepsToCome += Math.abs(previousTask - request.requestedFloor);
        } else {
            elevator.tasksList.some((task, index) => {
                let nextTask = elevator.tasksList[index] ? elevator.tasksList[index] : floor

                if (direction == 'up' && floor >= previousTask && floor <= nextTask) {
                    return stepsToCome += Math.abs(previousTask - floor)
                } else if (direction == 'down' && floor <= previousTask && floor >= nextTask) {
                    return stepsToCome += Math.abs(previousTask - floor)
                } else stepsToCome += Math.abs(previousTask - nextTask)

                previousTask = nextTask
            })
        }



        return stepsToCome
    }
}

class ElevatorRequest {
    constructor(floor, direction) {
        this.requestedFloor = floor
        this.direction = direction
        this.createdAt = Date.now()
    }
}










// **************************************************************************

//                           FLOOR_BUTTON

// **************************************************************************

class FloorButton {
    constructor(direction, number, columnId) {
        this.columnId = columnId
        this.floorNumber = number
        this.direction = direction
        this.columnId = columnId
        this.buttonsLight = false
    }

    requestElevator = () => {
        requestElevator(this.floorNumber, this.direction)
        this.toggleLight();
    }

    toggleLight = () => {
        this.buttonsLight = !this.buttonsLight
        console.log("FloorButton", this)
    }
}








// // **************************************************************************

// //    ELEVATOR

// // **************************************************************************

class Elevator {
    constructor(elevator, elevatorId) {
        this.id = elevatorId
        this.currentFloor = elevator.currentFloor ? elevator.currentFloor : 1
        this.doorState = elevator.doorState ? elevator.doorState : 'closed'
        this.isSafe = elevator.isSafe ? elevator.isSafe : 'true'
        this.maxWeight = elevator.maxWeight ? elevator.maxWeight : 3500
        this.buttonsList = elevator.buttonsList ? elevator.buttonsList : []
        this.tasksList = elevator.tasksList ? elevator.tasksList : []
    }


    getDirection = () => {
        let movingDown = (this.tasksList.length == 1 && this.currentFloor > this.tasksList[0]) ||
            this.tasksList[this.tasksList.length - 2] > this.tasksList[this.tasksList.length - 1]

        let movingUp = (this.tasksList.length == 1 && this.currentFloor < this.tasksList[0]) ||
            this.tasksList[this.tasksList.length - 2] < this.tasksList[this.tasksList.length - 1]


        if (this.tasksList.length == 0) return null
        else if (movingDown) {
            return 'down'
        } else if (movingUp) {
            return 'up'
        }
    }

    isComming = (request) => {
        let elevatorDirection = this.getDirection()
        let isComming = false

        if (elevatorDirection == request.direction) {
            if (elevatorDirection == "up") {
                isComming = this.tasksList.some(task => {
                    if (task <= request.requestedFloor) return true
                })
            } else if (elevatorDirection == "down") {
                isComming = this.tasksList.some(task => {
                    if (task >= request.requestedFloor) return true
                })
            }
        }

        return isComming
    }

    operate = () => {
        let again = true
        while (again) {
            if (this.tasksList[0] > this.currentFloor){
                 this.moveUp()
            }
            else if (this.tasksList[0] < this.currentFloor) {
                this.moveDown()
            }
            else {
                console.log('you are at your destination: ', this.currentFloor)
                this.openDoor()
                this.removeTask(this.tasksList[0])
                again = false

                setTimeout(()=>{
                   this.closeDoor()
                }, 5000)
            }   
        }
    }

    moveUp = () => {
        this.currentFloor = this.currentFloor + 1
        console.log('moving up to floor: ', this.currentFloor)
    }

    moveDown = () => {
        this.currentFloor = this.currentFloor - 1
        console.log('moving down to floor: ', this.currentFloor)
    }

    openDoor = () => {
        this.doorState == 'opened'
        this.removeTask()
        console.log('door is now open')
    }

    closeDoor = () => {
        this.doorState == "closed"
        console.log('door is closed')
        if (this.tasksList.length > 0) this.operate()
    }

    addTask = (task) => {    
        this.tasksList.push(task)
        console.log(task)

        this.operate()
    }

    removeTask = (task) => {
        this.tasksList.shift()
    }
}














const numberOfFloors = 10
const numberOfColumns = 1
const battery = new Battery(numberOfFloors, numberOfColumns);

console.log(`
**********************************************************************

    TO PUSH A BUTTON SIMPLY CALL requestElevator(floor, direction)
    direction == 'up' || 'down'

**********************************************************************
`)

const requestElevator = (floor, direction) => {
    let query = battery.getColumnButton(floor, direction)
    console.log(query)


    query ? query.column.requestElevator(floor, direction) : console.log('the request you trying to make is not possible because only God can')
}

const requestFloor = (elevator, requestedFloor) => {
    battery.columns.forEach(col => {
        col.elevators.some(elevatorItem => {
            if (elevatorItem.id == elevator) {
                elevatorItem.addTask(requestedFloor)
                console.log('Elevator Found, ', elevatorItem)
                return true
            }
        })
    })

}



























// const battery = {
//     numberOfFloors: 60,
//     numberOfColumns: 4,
// }

// const elevatorsList = [{
//     column: 0,
//     state: 'idle',
//     currentFloor: 2,
//     doorState: 'closed',
//     isSafe: true,
//     buttonsList: [],
//     tasksList: []
// }, {
//     column: 0,
//     state: 'moving',
//     currentFloor: 10,
//     doorState: 'closed',
//     isSafe: true,
//     buttonsList: [],
//     tasksList: []
// }]





// // **************************************************************************

// //    BATTERY

// // **************************************************************************


// class Battery {
//     constructor(numberOfFloors, numberOfColumns) {
//         this.columns = this.initColumns(numberOfFloors, numberOfColumns)

//         console.log(`[Battery] | constructor using numberOfFLoor: ${numberOfFloors} and numberOfColumns: ${numberOfColumns}`)
//     }

//     initColumns = (numberOfFLoors, numberOfColumns) => {
//         let range = Math.floor(numberOfFLoors / numberOfColumns);
//         let columns = []

//         let floor = 1
//         let maxFloor = numberOfFLoors - floor
//         for(; floor <= maxFloor; floor += range){
//             let column = {
//                 lowestFloorLevel: floor ,
//                 highestFloorLevel: floor + range                
//             }

//             columns.push(new Column(Math.random, column));
//             console.log(`[initColumns] | range: ${range}, floor: ${floor}`)
//         }

//         return columns
//     };


//     RequestElevator = (RequestedFloor, Direction, ) => {
//         let request = new Request(RequestedFloor, Direction);
//         console.log('floor requested');
//     }
// }





// const elevators = [

// ]





// // **************************************************************************

// //    COLUMN

// // **************************************************************************

// class Column {
//     constructor(id, column) {
//         this.id = id;
//         this.floorsList = this.initFloors(column.lowestFloorLevel, column.highestFloorLevel)
//         //this.elevatorsList = this.initElevators(column.elevatorsList)
//         console.log(`new Column ${this.id}`)
//     }

//     initFloors = (lowest, highest) => {
//         let floors = [];
//         floors.push(new Floor(0, this.id));

//         let i = lowest
//         while (i <= highest) {
//             floors.push(new Floor(i, this.id))
//             i++
//         }


//         return floors
//     }

//     initElevators = (elevatorsList) => {
//         let elevators = [];

//         let i = 1
//         const iMax = elevatorsList.length
//         for (; i < iMax; i++) {
//             elevators.push(new Elevator(elevatorsList[i]));
//         }

//         return elevators
//     }

// }










// // **************************************************************************

// //    FLOOR

// // **************************************************************************

// class Floor {
//     constructor(floorNumber, columnId) {
//         this.floorNumber = floorNumber;
//         this.floorButtons = this.initFloorButtons(this.floorNumber, columnId)
//         this.columnId = columnId
//         this.requestsList = []

//         console.log(`new Floor ${floorNumber} || on column ${this.columnId}`)
//     }

//     initFloorButtons = (floorNumber, columnId) => {
//         const buttons = []

//         if (floorNumber != battery.numberOfFloors) buttons.push(new FloorButton('up', floorNumber, columnId))
//         if (this.floorNumber != 0) buttons.push(new FloorButton('down', floorNumber, columnId));

//         return buttons
//     }
// }










// // **************************************************************************

// //    FLOOR BUTTON

// // **************************************************************************

// class FloorButton {
//     constructor(direction, number, columnId) {
//         this.floorNumber = number
//         this.direction = direction
//         this.columnId = columnId
//         this.buttonsLight = false

//         console.log(`new FloorButton | direction: ${this.direction}, floorNumber: ${this.floorNumber}, column: ${this.columnId}`)
//     }

//     requestElevator = () => {
//         // let request = new ElevatorRequest(this.floorNumber, this.direction, this.columnId)

//         // this.toggleLight();
//         // this.request = request
//     }

//     toggleLight = () => {
//         this.buttonsLight = !this.buttonsLight
//     }
// }


// class Request {
//     constructor(requestedFloor, direction, columnId) {
//         this.requestedFloor = requestElevator
//         this.direction = direction
//         this.createdAt = Date.now()
//         this.arrivedAt = null
//         this.columnId = columnId
//     }
// }


























// const b1 = new Battery(battery.numberOfFloors, battery.numberOfColumns)